docs/examples/cpp.md

```diff
@@ -0,0 +1,115 @@
+# C++ example
+
+In this example, we will walk through a simple C++ app that uses CMake.
+
+Let's assume our code is structured as follows:
+```
+.
+├── build.earth
+└── src
+    ├── CMakeLists.txt
+    ├── fib.cpp
+    ├── fib.h
+    └── main.cpp
+```
+
+Our program will be split between two different cpp files; the main.cpp file:
+
+```C++
+#include <iostream>
+
+#include "fib.h"
+
+int main(int argc, char *argv[]) {
+	for( int i = 0; i < 5; i++ ) {
+		std::cout << "fib(" << i << ") = " << fib(i) << std::endl;
+	}
+	return 0;
+}
+```
+
+and a file containing our fibonacci function:
+
+```C++
+#include "fib.h"
+
+int fib(int n)
+{
+	if( n <= 0 ) {
+		return 0;
+	}
+	if( n == 1 ) {
+		return 1;
+	}
+	return fib(n-1) + fib(n-2);
+}
+```
+
+We will use CMake to manage the build process of the c++ code, with the following CMakeList.txt file:
+
+```
+cmake_minimum_required(VERSION 2.8.9)
+project (fibonacci)
+add_executable(fibonacci main.cpp fib.cpp)
+```
+
+Cmake caches object files under CmakeFiles which allows cmake to only recompile objects when the corresponding
+source code changes. We will use a mount-based cache to cache these temporary files to allow for faster builds
+on a local machine. Here's a sample `earth.build` file:
+
+```Dockerfile
+# build.earth
+FROM ubuntu:20.10
+
+# configure apt to be noninteractive
+ENV DEBIAN_FRONTEND noninteractive
+ENV DEBCONF_NONINTERACTIVE_SEEN true
+
+# install dependencies
+RUN apt-get update && apt-get install -y build-essential cmake
+
+WORKDIR /code
+
+code:
+  COPY src src
+  SAVE IMAGE
+
+build:
+  FROM +code
+  RUN cmake src
+  # cache cmake temp files to prevent rebuilding .o files when the .cpp files don't change
```

Comment:
Might make sense to split this comment over two lines, so it's easier to read without side-scrolling.